### PR TITLE
Add MathCAT as an add-on

### PR DESCRIPTION
**Pull request by NSoiffer**
### Release information
- Name (Example, Add-on Updater): MathCAT
- Author (example, Your Name): Neil Soiffer
- Repo (example, https://github.com/yourname/addonName): https://github.com/NSoiffer/MathCATForPython
- Version (example, 21.07): 0.1.13
- Update channel (example, stable, dev or stable and dev): stable
- NVDA compatibility (example, 2021.1 and beyond): 2022.1

### Additional information
MathCAT is designed to eventually replace MathPlayer because MathPlayer is no longer supported. MathCAT generates speech and braille from MathML. The speech is enhanced with prosody so that it sounds more natural. The speech can be navigated in three modes using the same commands as MathPlayer. In addition, the navigation node is indicated on a braille display. Both Nemeth and UEB technical are supported.

MathCAT adds a settings menu to NVDA's preferences menu. In the settings menu, numerous options in MathCAT can be set to control the speech, navigation, and braille.

### CODE REVIEW

**Comment by XLTechie** [comment]
> MathCAT adds a settings menu to NVDA's tools menu.

May I ask why you are placing the settings on the Tools menu, instead of in NVDA 
settings panels, where most users of add-ons expect to find settings?

**Comment by NSoiffer** [comment]
I copied what Access8Math does. If that's not appropriate, can you point me
to an addon that uses the settings menu (which I agree makes more sense).
Much easier to copy/modify what someone else did than try to figure it out
in a void.

**Comment by josephsl** [comment]
Hi, StationPlaylist add-on is a good example of an add-on that adds a single menu entry to preferences menu, and if the add-on settings can fit inside a settings panel, we have plenty of examples 9Add-on Updater is one). Thanks.

**Comment by XLTechie** [comment]
> I copied what Access8Math does. If that's not appropriate, can you point me

It's an add-on. Anything you want to do and users will put up with, is 
"appropriate". It just isn't common any more.

> Much easier to copy/modify what someone else did than try to figure it out
> in a void.

Speech Logger (has text fields, a folder browser, and a combobox):
https://addons.nvda-project.org/files/get.php?file=splogger

Add-on Updater (has checkboxes): 
https://addons.nvda-project.org/files/get.php?file=nvda3208

**Comment by XLTechie** [comment]
Your best bet might be to put it on the preferences menu for this release, as 
@josephsl suggests.
Doing that, you would likely only have to move where the menu entry shows up 
(from Tools to Preferences), which should be simple.

You could then work on the possibly more involved process of converting it to a 
settings panel for some future release, as in the example add-ons I gave, if 
you wanted to go that route.

SPL, that Joseph mentioned, is here:
https://addons.nvda-project.org/./addons/stationPlaylist.en.html

Place Markers uses Preferences as well, but with a submenu, if you need that:
https://addons.nvda-project.org/./addons/placeMarkers.en.html

**Comment by NSoiffer** [comment]
I'll go with Joseph's suggestion. MathCAT has a lot of options. So many that I needed to divide them into three categories (Speech, Navigation, Braille). I could create three entries in the Settings dialog, but that's more work, so I'll stick with simple for now. Thanks for everyone's suggestions -- I'm still learning what's the expected thing for NVDA.

I'll send a new message when I'm done, but the pointer will remain the same. No one has used this build yet.

**Comment by NSoiffer** [comment]
The MathCAT settings is moved to preferences. I updated the initial comment to say that so that it reflects the release state in case someone doesn't bother reading the whole thread.

**Comment by nvdaes** [comment]
Hello:
You can also create different groups in the same settings panel. You can see examples in NVDA, for example in the Document formatting panel, and in add-ons, for example, in cursorLocator:

https://github.com/nvdaes/cursorLocator

Let us know when this PR is ready to be reviewed

**Comment by NSoiffer** [comment]
I'm going to skip breaking it up for this version. The UI is generated from a tool, so breaking it apart isn't trivial. The add-on is ready for review.

A few things to note:
1. The guts of the add-on is written in Rust. It is compiled for 32-bit Python 3.7. If you want to look at the source code, that is in [my github repo](https://github.com/NSoiffer/MathCAT). My hope is that MathCAT is used by several screen readers and other AT. I hope to hook up with Orca (Linux) this summer or fall. They don't have an add-on architecture, so I need to get buy-in from the Orca developers and better math support seems not to be a priority at the moment... or at least I haven't talked to the right developer yet. 
2. The speech and braille are mostly done in external rules and they show up in the Rules dir when add-on is installed.
3. There is python glue that does the work of connecting MathCAT to NVDA.  The `__init__.py` and`MathCAT.py` files are the main files for that. The other two `py` files are for the interface.

Question: once approved, what's the proper channel for announcing the add-on to NVDA users?

**Comment by NSoiffer** [comment]
I'm going to skip breaking it up for this version. The UI is generated from a tool, so breaking it apart isn't trivial. The add-on is ready for review.

A few things to note:
1. The guts of the add-on is written in Rust. It is compiled for 32-bit Python 3.7. If you want to look at the source code, that is in [my github repo](https://github.com/NSoiffer/MathCAT). My hope is that MathCAT is used by several screen readers and other AT. I hope to hook up with Orca (Linux) this summer or fall. They don't have an add-on architecture, so I need to get buy-in from the Orca developers and better math support seems not to be a priority at the moment... or at least I haven't talked to the right developer yet. 
2. The speech and braille are mostly done in external rules and they show up in the Rules dir when add-on is installed.
3. There is python glue that does the work of connecting MathCAT to NVDA.  The `__init__.py` and`MathCAT.py` files are the main files for that. The other two `py` files are for the interface.

Question: once approved, what's the proper channel for announcing the add-on to NVDA users?

**Comment by XLTechie** [comment]
> Question: once approved, what's the proper channel for announcing the add-on to NVDA users?

Every channel you can think of where it might be appropriate.

https://groups.io/g/nvda
https://groups.io/g/nvda-addons

Being just two starting possibilities.

**Comment by nvdaes** [comment]
> Question: once approved, what's the proper channel for announcing the add-on to NVDA users?

In addition to @XLTechie response, remember that, if you want to have a webpage for this add-on on the website (not just a permanent link), you need to update the documentation in the translation repo:

https://github.com/nvaccess/nvda/wiki/Translating

You can checkout just the website folder if you want to update the documentation for your add-on (it's not needed to download the whole repo).

**Comment by NSoiffer** [comment]
I have updated MathCAT to deal with a security concern.

It also has some refinements to a couple of rules. Because of this, the version number has been bumped to v.0.1.14.

File: get.php [hunk 1]
```diff
@@ -72,6 +72,7 @@
 	"killnvda" => "https://github.com/tspivey/killNVDA/releases/download/v0.2/killNVDA-0.2.nvda-addon",
 	"lambda" => "https://github.com/lambda-nvda/lambdaNvda/releases/download/1.3.0/lambda-1.3.0.nvda-addon",
 	"lambda-dev" => "https://github.com/lambda-nvda/lambdaNvda/releases/download/1.3.0/lambda-1.3.0.nvda-addon",
+	"MathCAT" => "https://github.com/NSoiffer/MathCATForPython/releases/download/v.0.1.14/MathCAT-0.1.14.nvda-addon"
```
Comment (nvdaes) [inline get.php:75]:
```suggestion
	"MathCAT" => "https://github.com/NSoiffer/MathCATForPython/releases/download/v.0.1.14/MathCAT-0.1.14.nvda-addon",
```

**Comment by NSoiffer** [comment]
Good catch -- I've added the comma.

**Comment by XLTechie** [comment]
May I suggest that the get.php key not be written in mixed case? That is 
non-standard for get.php, I think.

**Comment by nvdaes** [comment]
Hi Luke:

Anyway there are other ids (keys) in get.php writen with mixec
capitalized characters. I'll approve this today.

2022-05-29 7:43 GMT+02:00, Luke Davis ***@***.***>:

**Comment by nvdaes** [comment]
> Good catch -- I've added the comma.

Seems the comma is not addedfor now.

**Comment by XLTechie** [comment]
Twas just a suggestion.

**Comment by NSoiffer** [comment]
Weird about the comma, I was sure I had done the commit. I guess just wishful thinking or confused with one of the other projects I'm working on. Committed now.

**Review by nvdaes — CHANGES_REQUESTED** [review]
Seems that your branch doesn't have last changes made on this repo. Please ensure it's updated.

**Comment by NSoiffer** [comment]
Something seems to be messed up with my fork because git is claiming everything is up to date with the upstream git (so it should have had the latest changes). I'm going to close this PR and start from scratch and hopefully it will be right the next time.

**Comment by nvdaes** [comment]
Hi Luke: Anyway there are several ids or keys using capitalized characters. I"ll approve this today.

Enviado desde mi iPhone

**Pull request closed without merging**